styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -109,54 +123,156 @@
+    this(client, stateManager, stats, serviceAccountKeyManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS);
   }
 
   @Override
   public String start(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
+    ensureSecrets(workflowInstance, runSpec);
     try {
-      runSpec.secret().ifPresent(specSecret -> {
-        final Secret secret = client.secrets().withName(specSecret.name()).get();
-        if (secret == null) {
-          LOG.error("[AUDIT] Workflow {} refers to a non-existent secret {}",
-                    workflowInstance.workflowId(), specSecret.name());
-          throw new InvalidExecutionException(
-              "Referenced secret '" + specSecret.name() + "' was not found");
-        } else {
-          LOG.info("[AUDIT] Workflow {} refers to secret {}",
-                   workflowInstance.workflowId(), specSecret.name());
-        }
-      });
-
       final Pod pod = client.pods().create(createPod(workflowInstance, runSpec));
       return pod.getMetadata().getName();
     } catch (KubernetesClientException kce) {
       throw new IOException("Failed to create Kubernetes pod", kce);
     }
   }
 
+  private void ensureSecrets(WorkflowInstance workflowInstance, RunSpec runSpec) {
+
+    if (runSpec.serviceAccount().isPresent()) {
+      final String secretName = buildSecretName(runSpec.serviceAccount().get());
+      final Secret secret = client.secrets().withName(secretName).get();
+      if (secret == null) {
+        final String serviceAccount = runSpec.serviceAccount().get();
+
+        final ServiceAccountKey jsonKey;
+        final ServiceAccountKey p12Key;
+        try {
+          jsonKey = serviceAccountKeyManager.createJsonKey(serviceAccount);
+          p12Key = serviceAccountKeyManager.createP12Key(serviceAccount);
+        } catch (IOException e) {
+          logger.error("Failed to create service account keys", e);
+          throw new RuntimeException(e);
+        }
+
+        final Map<String, String> keys = ImmutableMap.of(
+            STYX_WORKFLOW_SA_JSON_KEY, jsonKey.toString(),
+            STYX_MANAGED_WORKFLOW_SA_P12_KEY, p12Key.toString()
+        );
+
+        final Map<String, String> annotations = ImmutableMap.of(
+            STYX_WORKFLOW_SA_JSON_KEY_NAME_ANNOTATION, jsonKey.getName(),
+            STYX_WORKFLOW_SA_P12_KEY_NAME_ANNOTATION, p12Key.getName(),
+            STYX_WORKFLOW_SA_SECRET_ANNOTATION, serviceAccount
+        );
+
+        client.secrets().create(new SecretBuilder()
+                                    .withNewMetadata()
+                                    .withName(secretName)
+                                    .withAnnotations(annotations)
+                                    .endMetadata()
+                                    .withData(keys)
+                                    .build());
+        LOG.info("[AUDIT] Secret {} created for service account {}", secretName, serviceAccount);
+      }
+    }
+
+    if (runSpec.secret().isPresent()) {
```

Comment:
@bergman @fabriziodemaria @danielnorberg let's continue here. do we want them to co-eixst.